apps/api/src/app/rate-limiting/guards/throttler.guard.ts

```diff
@@ -0,0 +1,182 @@
+import {
+  InjectThrottlerOptions,
+  InjectThrottlerStorage,
+  ThrottlerException,
+  ThrottlerGuard,
+  ThrottlerModuleOptions,
+  ThrottlerOptions,
+  ThrottlerStorage,
+} from '@nestjs/throttler';
+import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from '@nestjs/common';
+import { EvaluateApiRateLimit, EvaluateApiRateLimitCommand } from '../usecases/evaluate-api-rate-limit';
+import { Reflector } from '@nestjs/core';
+import { FeatureFlagCommand, GetIsApiRateLimitingEnabled } from '@novu/application-generic';
+import { ApiRateLimitCategoryEnum, ApiRateLimitCostEnum, IJwtPayload } from '@novu/shared';
+import * as jwt from 'jsonwebtoken';
+import { ThrottlerCost, ThrottlerCategory } from './throttler.decorator';
+
+export enum RateLimitHeaderKeysEnum {
+  RATE_LIMIT_REMAINING = 'RateLimit-Remaining',
+  RATE_LIMIT_LIMIT = 'RateLimit-Limit',
+  RATE_LIMIT_RESET = 'RateLimit-Reset',
+  RATE_LIMIT_POLICY = 'RateLimit-Policy',
+  RETRY_AFTER = 'Retry-After',
+  USER_AGENT = 'User-Agent',
+  AUTHORIZATION = 'Authorization',
+}
+
+export const THROTTLED_EXCEPTION_MESSAGE = 'API rate limit exceeded';
+export const ALLOWED_AUTH_SCHEMES = ['ApiKey'];
+
+const defaultApiRateLimitCategory = ApiRateLimitCategoryEnum.GLOBAL;
+const defaultApiRateLimitCost = ApiRateLimitCostEnum.SINGLE;
+
+/**
+ * An interceptor is used instead of a guard to ensure that Auth context is available.
+ * This is currently necessary because we do not currently have a global guard configured for Auth,
+ * therefore the Auth context is not guaranteed to be available in the guard.
```

Comment:
What would it take to switch this to a guard? and should we?

Reply:
After making these changes, we can remove a considerable number of individual controller-level and method-level use of the AuthGuard, providing better maintainability.

Given the sizeable scope and many controllers/methods touched, I think we should make these change in a separate PR.

Reply:
I agree, Can you please create a set of tasks to do this in linear so that way I can prioritize this follow-up tasks?